curvefs/src/mds/topology/topology_metric.h

```diff
@@ -193,10 +201,26 @@ struct FsMetric {
     // inode number
 
     bvar::Status<uint64_t> inodeNum_;
+    using statusPtr = std::shared_ptr<bvar::Status<uint64_t>>;
+    std::unordered_map<FsFileType, statusPtr> fileType2InodeNum_;
 
     explicit FsMetric(FsIdType fsId)
         : inodeNum_(kTopologyFsMetricPrefix,
-                    std::to_string(fsId) + "_inode_num", 0) {}
+                    std::to_string(fsId) + "_inode_num", 0) {
+        for (int i = FsFileType_MIN; i <= FsFileType_MAX; ++i) {
+            if (FsFileType_IsValid(i)) {
+                auto inodeNumTmpPtr = std::make_shared<bvar::Status<uint64_t>>(
+                    kTopologyFsMetricPrefix,
+                    std::to_string(fsId) + "_" +
+                        FsFileType_Name(static_cast<FsFileType>(i)) +
+                        "_inode_num",
+                    0);
+
+                fileType2InodeNum_.emplace(static_cast<FsFileType>(i),
+                                           inodeNumTmpPtr);
```

Comment:
> std::move(inodeNumTmpPtr), and can we just store it by value?

fix

Reply:
> std::move(inodeNumTmpPtr), and can we just store it by value?


If so, it will fail at compile time.like this：

```
ERROR: /home/chengyi01/code/curve/curvefs/src/mds/topology/BUILD:19:11: Compiling curvefs/src/mds/topology/topology_metric.cpp failed: (Exit 1): gcc failed: error executing command /usr/local/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++0x' -MD -MF ... (remaining 99 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
curvefs/src/mds/topology/topology_metric.cpp: In lambda function:
curvefs/src/mds/topology/topology_metric.cpp:44:30: warning: unused parameter 'ms' [-Wunused-parameter]
         [](const MetaServer &ms) { return true; });
                              ^~
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/string:41,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/ext/new_allocator.h: In instantiation of 'void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >; _Args = {curvefs::metaserver::FsFileType, bvar::Status<long unsigned int, void>}; _Tp = std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >]':
/usr/include/c++/6/bits/alloc_traits.h:475:4:   required from 'static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*, _Args&& ...) [with _Up = std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >; _Args = {curvefs::metaserver::FsFileType, bvar::Status<long unsigned int, void>}; _Tp = std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> > >]'
/usr/include/c++/6/bits/hashtable_policy.h:1953:37:   required from 'std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_type* std::__detail::_Hashtable_alloc<_NodeAlloc>::_M_allocate_node(_Args&& ...) [with _Args = {curvefs::metaserver::FsFileType, bvar::Status<long unsigned int, void>}; _NodeAlloc = std::allocator<std::__detail::_Hash_node<std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >, false> >; std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_type = std::__detail::_Hash_node<std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >, false>]'
/usr/include/c++/6/bits/hashtable.h:1513:77:   required from 'std::pair<typename std::__detail::_Hashtable_base<_Key, _Value, _ExtractKey, _Equal, _H1, _H2, _Hash, _Traits>::iterator, bool> std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::_M_emplace(std::true_type, _Args&& ...) [with _Args = {curvefs::metaserver::FsFileType, bvar::Status<long unsigned int, void>}; _Key = curvefs::metaserver::FsFileType; _Value = std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >; _Alloc = std::allocator<std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> > >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<curvefs::metaserver::FsFileType>; _H1 = std::hash<curvefs::metaserver::FsFileType>; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<false, false, true>; typename std::__detail::_Hashtable_base<_Key, _Value, _ExtractKey, _Equal, _H1, _H2, _Hash, _Traits>::iterator = std::__detail::_Node_iterator<std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >, false, false>; std::true_type = std::integral_constant<bool, true>]'
/usr/include/c++/6/bits/hashtable.h:728:69:   required from 'std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__ireturn_type std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::emplace(_Args&& ...) [with _Args = {curvefs::metaserver::FsFileType, bvar::Status<long unsigned int, void>}; _Key = curvefs::metaserver::FsFileType; _Value = std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >; _Alloc = std::allocator<std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> > >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<curvefs::metaserver::FsFileType>; _H1 = std::hash<curvefs::metaserver::FsFileType>; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<false, false, true>; std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__ireturn_type = std::pair<std::__detail::_Node_iterator<std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >, false, false>, bool>]'
/usr/include/c++/6/bits/unordered_map.h:380:54:   required from 'std::pair<typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, std::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__detail::__is_noexcept_hash<_Key, _Hash> > >::value, false, true> >::iterator, bool> std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::emplace(_Args&& ...) [with _Args = {curvefs::metaserver::FsFileType, bvar::Status<long unsigned int, void>}; _Key = curvefs::metaserver::FsFileType; _Tp = bvar::Status<long unsigned int>; _Hash = std::hash<curvefs::metaserver::FsFileType>; _Pred = std::equal_to<curvefs::metaserver::FsFileType>; _Alloc = std::allocator<std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> > >; typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, std::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__detail::__is_noexcept_hash<_Key, _Hash> > >::value, false, true> >::iterator = std::__detail::_Node_iterator<std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >, false, false>]'
./curvefs/src/mds/topology/topology_metric.h:227:27:   required from here
/usr/include/c++/6/ext/new_allocator.h:120:4: error: no matching function for call to 'std::pair<const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >::pair(curvefs::metaserver::FsFileType, bvar::Status<long unsigned int>)'
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/bits/stl_pair.h:427:9: note: candidate: template<class ... _Args1, long unsigned int ..._Indexes1, class ... _Args2, long unsigned int ..._Indexes2> std::pair<_T1, _T2>::pair(std::tuple<_Args1 ...>&, std::tuple<_Args2 ...>&, std::_Index_tuple<_Indexes1 ...>, std::_Index_tuple<_Indexes2 ...>)
         pair(tuple<_Args1...>&, tuple<_Args2...>&,
         ^~~~
/usr/include/c++/6/bits/stl_pair.h:427:9: note:   template argument deduction/substitution failed:
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/string:41,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/ext/new_allocator.h:120:4: note:   mismatched types 'std::tuple<_Elements ...>' and 'curvefs::metaserver::FsFileType'
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/bits/stl_pair.h:360:9: note: candidate: template<class ... _Args1, class ... _Args2> std::pair<_T1, _T2>::pair(std::piecewise_construct_t, std::tuple<_Args1 ...>, std::tuple<_Args2 ...>)
         pair(piecewise_construct_t, tuple<_Args1...>, tuple<_Args2...>);
         ^~~~
/usr/include/c++/6/bits/stl_pair.h:360:9: note:   template argument deduction/substitution failed:
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/string:41,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/ext/new_allocator.h:120:4: note:   'bvar::Status<long unsigned int>' is not derived from 'std::tuple<_Elements ...>'
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/bits/stl_pair.h:355:21: note: candidate: template<class _U1, class _U2, typename std::enable_if<(std::_PCC<((! std::is_same<const curvefs::metaserver::FsFileType, _U1>::value) || (! std::is_same<bvar::Status<long unsigned int>, _U2>::value)), const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >::_MoveConstructiblePair<_U1, _U2>() && (! std::_PCC<((! std::is_same<const curvefs::metaserver::FsFileType, _U1>::value) || (! std::is_same<bvar::Status<long unsigned int>, _U2>::value)), const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >::_ImplicitlyMoveConvertiblePair<_U1, _U2>())), bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair(std::pair<_U1, _U2>&&)
  explicit constexpr pair(pair<_U1, _U2>&& __p)
                     ^~~~
/usr/include/c++/6/bits/stl_pair.h:355:21: note:   template argument deduction/substitution failed:
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/string:41,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/ext/new_allocator.h:120:4: note:   mismatched types 'std::pair<_T1, _T2>' and 'curvefs::metaserver::FsFileType'
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/bits/stl_pair.h:345:12: note: candidate: template<class _U1, class _U2, typename std::enable_if<(std::_PCC<((! std::is_same<const curvefs::metaserver::FsFileType, _U1>::value) || (! std::is_same<bvar::Status<long unsigned int>, _U2>::value)), const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >::_MoveConstructiblePair<_U1, _U2>() && std::_PCC<((! std::is_same<const curvefs::metaserver::FsFileType, _U1>::value) || (! std::is_same<bvar::Status<long unsigned int>, _U2>::value)), const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >::_ImplicitlyMoveConvertiblePair<_U1, _U2>()), bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair(std::pair<_U1, _U2>&&)
  constexpr pair(pair<_U1, _U2>&& __p)
            ^~~~
/usr/include/c++/6/bits/stl_pair.h:345:12: note:   template argument deduction/substitution failed:
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/string:41,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/ext/new_allocator.h:120:4: note:   mismatched types 'std::pair<_T1, _T2>' and 'curvefs::metaserver::FsFileType'
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/bits/stl_pair.h:335:21: note: candidate: template<class _U1, class _U2, typename std::enable_if<(_MoveConstructiblePair<_U1, _U2>() && (! _ImplicitlyMoveConvertiblePair<_U1, _U2>())), bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair(_U1&&, _U2&&)
  explicit constexpr pair(_U1&& __x, _U2&& __y)
                     ^~~~
/usr/include/c++/6/bits/stl_pair.h:335:21: note:   template argument deduction/substitution failed:
/usr/include/c++/6/bits/stl_pair.h:334:38: error: no type named 'type' in 'struct std::enable_if<false, bool>'
                          bool>::type=false>
                                      ^~~~~
/usr/include/c++/6/bits/stl_pair.h:334:38: note: invalid template non-type parameter
/usr/include/c++/6/bits/stl_pair.h:326:12: note: candidate: template<class _U1, class _U2, typename std::enable_if<(_MoveConstructiblePair<_U1, _U2>() && _ImplicitlyMoveConvertiblePair<_U1, _U2>()), bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair(_U1&&, _U2&&)
  constexpr pair(_U1&& __x, _U2&& __y)
            ^~~~
/usr/include/c++/6/bits/stl_pair.h:326:12: note:   template argument deduction/substitution failed:
/usr/include/c++/6/bits/stl_pair.h:325:38: error: no type named 'type' in 'struct std::enable_if<false, bool>'
                          bool>::type=true>
                                      ^~~~
/usr/include/c++/6/bits/stl_pair.h:325:38: note: invalid template non-type parameter
/usr/include/c++/6/bits/stl_pair.h:317:17: note: candidate: template<class _U2, typename std::enable_if<_CopyMovePair<false, const curvefs::metaserver::FsFileType, _U2>(), bool>::type <anonymous> > std::pair<_T1, _T2>::pair(const _T1&, _U2&&)
        explicit pair(const _T1& __x, _U2&& __y)
                 ^~~~
/usr/include/c++/6/bits/stl_pair.h:317:17: note:   template argument deduction/substitution failed:
/usr/include/c++/6/bits/stl_pair.h:316:38: error: no type named 'type' in 'struct std::enable_if<false, bool>'
                          bool>::type=false>
                                      ^~~~~
/usr/include/c++/6/bits/stl_pair.h:316:38: note: invalid template non-type parameter
/usr/include/c++/6/bits/stl_pair.h:310:18: note: candidate: template<class _U2, typename std::enable_if<_CopyMovePair<true, const curvefs::metaserver::FsFileType, _U2>(), bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair(const _T1&, _U2&&)
        constexpr pair(const _T1& __x, _U2&& __y)
                  ^~~~
/usr/include/c++/6/bits/stl_pair.h:310:18: note:   template argument deduction/substitution failed:
/usr/include/c++/6/bits/stl_pair.h:309:38: error: no type named 'type' in 'struct std::enable_if<false, bool>'
                          bool>::type=true>
                                      ^~~~
/usr/include/c++/6/bits/stl_pair.h:309:38: note: invalid template non-type parameter
/usr/include/c++/6/bits/stl_pair.h:303:27: note: candidate: template<class _U1, typename std::enable_if<_MoveCopyPair<false, _U1, bvar::Status<long unsigned int, void> >(), bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair(_U1&&, const _T2&)
        explicit constexpr pair(_U1&& __x, const _T2& __y)
                           ^~~~
/usr/include/c++/6/bits/stl_pair.h:303:27: note:   template argument deduction/substitution failed:
/usr/include/c++/6/bits/stl_pair.h:302:38: error: no type named 'type' in 'struct std::enable_if<false, bool>'
                          bool>::type=false>
                                      ^~~~~
/usr/include/c++/6/bits/stl_pair.h:302:38: note: invalid template non-type parameter
/usr/include/c++/6/bits/stl_pair.h:296:18: note: candidate: template<class _U1, typename std::enable_if<_MoveCopyPair<true, _U1, bvar::Status<long unsigned int, void> >(), bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair(_U1&&, const _T2&)
        constexpr pair(_U1&& __x, const _T2& __y)
                  ^~~~
/usr/include/c++/6/bits/stl_pair.h:296:18: note:   template argument deduction/substitution failed:
/usr/include/c++/6/bits/stl_pair.h:295:38: error: no type named 'type' in 'struct std::enable_if<false, bool>'
                          bool>::type=true>
                                      ^~~~
/usr/include/c++/6/bits/stl_pair.h:295:38: note: invalid template non-type parameter
/usr/include/c++/6/bits/stl_pair.h:285:21: note: candidate: template<class _U1, class _U2, typename std::enable_if<(std::_PCC<((! std::is_same<const curvefs::metaserver::FsFileType, _U1>::value) || (! std::is_same<bvar::Status<long unsigned int>, _U2>::value)), const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >::_ConstructiblePair<_U1, _U2>() && (! std::_PCC<((! std::is_same<const curvefs::metaserver::FsFileType, _U1>::value) || (! std::is_same<bvar::Status<long unsigned int>, _U2>::value)), const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >::_ImplicitlyConvertiblePair<_U1, _U2>())), bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair(const std::pair<_U1, _U2>&)
  explicit constexpr pair(const pair<_U1, _U2>& __p)
                     ^~~~
/usr/include/c++/6/bits/stl_pair.h:285:21: note:   template argument deduction/substitution failed:
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/string:41,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/ext/new_allocator.h:120:4: note:   mismatched types 'const std::pair<_T1, _T2>' and 'curvefs::metaserver::FsFileType'
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/bits/stl_pair.h:276:19: note: candidate: template<class _U1, class _U2, typename std::enable_if<(std::_PCC<((! std::is_same<const curvefs::metaserver::FsFileType, _U1>::value) || (! std::is_same<bvar::Status<long unsigned int>, _U2>::value)), const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >::_ConstructiblePair<_U1, _U2>() && std::_PCC<((! std::is_same<const curvefs::metaserver::FsFileType, _U1>::value) || (! std::is_same<bvar::Status<long unsigned int>, _U2>::value)), const curvefs::metaserver::FsFileType, bvar::Status<long unsigned int> >::_ImplicitlyConvertiblePair<_U1, _U2>()), bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair(const std::pair<_U1, _U2>&)
         constexpr pair(const pair<_U1, _U2>& __p)
                   ^~~~
/usr/include/c++/6/bits/stl_pair.h:276:19: note:   template argument deduction/substitution failed:
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/string:41,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/ext/new_allocator.h:120:4: note:   mismatched types 'const std::pair<_T1, _T2>' and 'curvefs::metaserver::FsFileType'
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/bits/stl_pair.h:254:26: note: candidate: template<class _U1, class _U2, typename std::enable_if<(_ConstructiblePair<_U1, _U2>() && (! _ImplicitlyConvertiblePair<_U1, _U2>())), bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair(const _T1&, const _T2&)
       explicit constexpr pair(const _T1& __a, const _T2& __b)
                          ^~~~
/usr/include/c++/6/bits/stl_pair.h:254:26: note:   template argument deduction/substitution failed:
/usr/include/c++/6/bits/stl_pair.h:253:38: error: no type named 'type' in 'struct std::enable_if<false, bool>'
                          bool>::type=false>
                                      ^~~~~
/usr/include/c++/6/bits/stl_pair.h:253:38: note: invalid template non-type parameter
/usr/include/c++/6/bits/stl_pair.h:245:17: note: candidate: template<class _U1, class _U2, typename std::enable_if<(_ConstructiblePair<_U1, _U2>() && _ImplicitlyConvertiblePair<_U1, _U2>()), bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair(const _T1&, const _T2&)
       constexpr pair(const _T1& __a, const _T2& __b)
                 ^~~~
/usr/include/c++/6/bits/stl_pair.h:245:17: note:   template argument deduction/substitution failed:
/usr/include/c++/6/bits/stl_pair.h:244:38: error: no type named 'type' in 'struct std::enable_if<false, bool>'
                          bool>::type=true>
                                      ^~~~
/usr/include/c++/6/bits/stl_pair.h:244:38: note: invalid template non-type parameter
/usr/include/c++/6/bits/stl_pair.h:227:26: note: candidate: template<class _U1, class _U2, typename std::enable_if<std::__and_<std::is_default_constructible<_Tp>, std::is_default_constructible<_U2>, std::__not_<std::__and_<std::__is_implicitly_default_constructible<_U1>, std::__is_implicitly_default_constructible<_U2> > > >::value, bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair()
       explicit constexpr pair()
                          ^~~~
/usr/include/c++/6/bits/stl_pair.h:227:26: note:   template argument deduction/substitution failed:
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/string:41,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/ext/new_allocator.h:120:4: note:   candidate expects 0 arguments, 2 provided
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/bits/stl_pair.h:214:26: note: candidate: template<class _U1, class _U2, typename std::enable_if<std::__and_<std::__is_implicitly_default_constructible<_U1>, std::__is_implicitly_default_constructible<_U2> >::value, bool>::type <anonymous> > constexpr std::pair<_T1, _T2>::pair()
       _GLIBCXX_CONSTEXPR pair()
                          ^~~~
/usr/include/c++/6/bits/stl_pair.h:214:26: note:   template argument deduction/substitution failed:
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/string:41,
                 from external/com_github_apache_brpc/src/butil/logging.h:26,
                 from external/com_github_apache_brpc/src/bvar/reducer.h:22,
                 from external/com_github_apache_brpc/src/bvar/bvar.h:21,
                 from ./curvefs/src/mds/topology/topology_metric.h:26,
                 from curvefs/src/mds/topology/topology_metric.cpp:23:
/usr/include/c++/6/ext/new_allocator.h:120:4: note:   candidate expects 0 arguments, 2 provided
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```